### ISSUE TITLE
Pdf files created lost their fillables forms 

### ISSUE THREAD

**Issue by odadoun** [pikepdf/pikepdf#98]
Hello,
I have a fillable document that I want to split. I have used this code:
[pdf_ori=Pdf.open("DOC-sujet.pdf")
for n, page in enumerate(pdf_ori.pages):
     dst = Pdf.new()
     dst.pages.append(page)
     dst.save('{:02d}.pdf'.format(n))](url)
But all the pdf produced are not anymore fillable under Okular linux pdf preview, but it works with Aperçu from OS X. Do I need to add some header ?
The other way is to remove the pages from the initial document, but is not convenient.
By advance thank you
Olivier

**Comment by jbarlow83** [comment]
Please add the PDF you're working with.

**Comment by odadoun** [comment]
[DOC-sujet.pdf](https://github.com/pikepdf/pikepdf/files/4505706/DOC-sujet.pdf)

**Comment by jbarlow83** [comment]
Some of the information pertaining to the fillable form is "global" to the PDF rather than attached to the page. Specifically, the information in the `pdf_ori.Root.AcroForm` needs to be reproduced in the split documents. 

For a splitting short name of pages like this you can probably just remove all pages but the one to keep, and use `pdf_ori.remove_unreferenced_resources()` to clean up. Less efficient, but all less coding effort.

The behavior of those two different PDF viewers is explained by differences in their ability to cope with incomplete information in the generated PDF - much like web browsers trying to parse a nonconforming HTML document.

**Comment by odadoun** [comment]
Thanks for your prompt answer.  In fact I have a huge document of 400 pages to split in 100 pdf :-) Is there a way to reproduce pdf_ori.Root.AcroForm in my document ? If it is to complex I will try your solution by removing not intended pages

**Comment by jbarlow83** [comment]
Yes, you can copy the necessary portions. Roughly speaking you will require all portions of that structure that are also referenced by the page you are copying, and you'll make indirect references with `Pdf.make_indirect`.

pikepdf allows you to make all of the necessary low-level manipulations, but it does not understand the higher level semantics of the AcroForm data structure. This is true for many PDF features.

I recommend reading the PDF 1.7 Reference Manual on AcroForm to see what is required.

**Comment by odadoun** [comment]
Sorry but I haven't succeeded in producing my documents ... 
pdf_ori=Pdf.open("sujet.pdf")
pdf_temp=pdf_ori
del pdf_temp.pages[2:4]
temp_name="output-1-2.pdf"
pdf_temp.save(temp_name)
pdf_temp.remove_unreferenced_resources()
pdf_temp=pdf_ori
del pdf_temp.pages[1:3]
temp_name="output-3-4.pdf"
pdf_temp.save(temp_name)

pdf_ori.remove_unreferenced_resources() or pdf_temp.remove_unreferenced_resources() seems to have no effect ... any idea ? by advance thank you

**Comment by jbarlow83** [comment]
pdf_temp.remove_unreferenced_resources() must be called before saving, not after.

You will also need to re-open the original file each time, for example

```python
with Pdf.open('sujet.pdf') as pdf:
    count = len(pdf.pages)
for n in range(count):
    with Pdf.open('sujet.pdf') as pdf:
        del pdf.pages[0:n]
        del pdf.pages[n+1:]
        pdf.remove_unreferenced_resources()
        pdf.save(f'sujet-{n}.pdf')
```

**Comment by jbarlow83** [comment]
Addressed in the upcoming pikepdf v10.9.

The split loses the fillable fields because a naive `pages.append()` / `pages.extend()` across documents copies the page but not the associated AcroForm form fields and their widget annotations, so viewers like Okular no longer treat the result as a form.

v10.9 adds `Pdf.add_pages_from()`, which copies pages between documents while preserving interactive AcroForm form fields:

```python
src = Pdf.open('DOC-sujet.pdf')
for n, page in enumerate(src.pages):
    dst = Pdf.new()
    dst.add_pages_from(src, [n])
    dst.save(f'{n:02d}.pdf')
```

As part of the same change, naive cross-document `pages.extend()` / `append()` that orphans form widgets now emits a `FormCopyWarning` to point you at `add_pages_from()`. Closing as resolved in v10.9.
